nixos/modules/services/misc/anki-sync-server.nix

```diff
@@ -0,0 +1,153 @@
+      "\n"
+      (x: ''export SYNC_USER${toString x.i}=${strings.escapeShellArg x.user.username}:${strings.escapeShellArg x.user.password}'')
+      usersWithIndexesNoFile
+    }
+    exec ${cfg.package}/bin/anki --syncserver
+  '';
+in {
+  options.services.anki-sync-server = {
+    enable = mkEnableOption (lib.mdDoc "anki-sync-server");
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.anki-bin;
+      defaultText = literalExpression "pkgs.anki-bin";
+      description = lib.mdDoc "The package to use for the anki command.";
+    };
+
+    host = mkOption {
+      type = types.str;
+      default = "localhost";
+      description = lib.mdDoc "anki-sync-server host";
+    };
+
+    port = mkOption {
+      type = types.port;
+      default = 27701;
+      description = lib.mdDoc "anki-sync-server port";
+    };
+
+    openFirewall = mkOption {
+      default = false;
+      type = types.bool;
+      description = lib.mdDoc "Whether to open the firewall for the specified port.";
+    };
+
+    users = mkOption {
+      type = with types;
+        listOf (submodule {
+          options = {
+            username = mkOption {
+              type = str;
+              description = lib.mdDoc "User name accepted by anki-sync-server.";
+            };
+            password = mkOption {
+              type = nullOr str;
+              default = null;
+              description = lib.mdDoc ''
+                Password accepted by anki-sync-server for the associated username.
+                **WARNING**: This option is **not secure**. This password will
+                be stored in *plaintext* and will be visible to *all users*.
+                See {option}`services.anki-sync-server.users.passwordFile` for
+                a more secure option.
+              '';
+            };
+            passwordFile = mkOption {
+              type = nullOr path;
+              default = null;
+              description = lib.mdDoc ''
+                File containing the password accepted by anki-sync-server for
+                the associated username.  Make sure to make readable only by
+                root.
+              '';
+            };
+          };
+        });
+      description = lib.mdDoc "List of user-password pairs to provide to the sync server.";
+    };
+  };
+
+  config = mkIf cfg.enable {
+    assertions = [
+      {
+        assertion = (builtins.length usersWithIndexesFile) + (builtins.length usersWithIndexesNoFile) > 0;
+        message = "At least one username-password pair must be set.";
+      }
+      {
+        assertion =
+          (
+            builtins.compareVersions
+            (elemAt (builtins.match ".*-(([0-9]+\\.)*[0-9]+)$" cfg.package.name) 0)
+            "2.1.66"
+          )
+          != -1;
+        message = "anki package <2.1.66 not supported by this module. Earliest supported commit: https://github.com/NixOS/nixpkgs/commit/05727304f8815825565c944d012f20a9a096838a";
```

Comment:
(does this make sense in nixpkgs? To avoid this commit being backported? my impression was that new modules aren't usually backported to stable trees)

Reply:
@de11n Do you think I should take this out because the version in master nixpkgs works? I put this in originally because I found it useful on my machine.

Reply:
I don't know the answer to this. I think it might depend on how common/likely we think it will be for users to configure their own version of `anki`. If there's some reason to believe that will be common, then it makes sense to help users get it right. If we expect folks to rarely, if ever, configure this manually, then it's probably overkill. NixOS tests would be ideal to catch breakages.

Reply:
Hmm, this flat out didn't work for me:
```
error:
       … while evaluating the attribute 'drvPath'
         at /home/user/nixpkgs/lib/customisation.nix:222:7:
          221|     in commonAttrs // {
          222|       drvPath = assert condition; drv.drvPath;
             |       ^
          223|       outPath = assert condition; drv.outPath;
       … while calling the 'derivationStrict' builtin
         at /builtin/derivation.nix:9:12: (source not available)
       (stack trace truncated; use '--show-trace' to show the full trace)
       error: value is null while a list was expected
```
Very helpful error and --show-trace didn't help as usual; sometimes nix is just bad at this... I've flat out removed the check in my test branch:
https://github.com/martinetd/nixpkgs/commit/bd6a4753f10ce96ea61db4417cddc8a1b7212b0b
but feel free to amend your commits

Reply:
We removed this warning, because upstream is now using Anki 23.10.